Lib/test/test_import/__init__.py

```diff
@@ -111,6 +111,27 @@ def test_from_import_missing_attr_path_is_canonical(self):
         self.assertIn(cm.exception.name, {'posixpath', 'ntpath'})
         self.assertIsNotNone(cm.exception)
 
+    def test_from_import_star_invalid_type(self):
+        TESTFNnoat = TESTFN[1:]
+        source = TESTFNnoat + os.extsep + "py"
+        sys.path.insert(0, os.curdir)
+        try:
+            with open(source, "w") as f:
+                f.write("__all__ = [1]")
+            with self.assertRaisesRegex(
+                    TypeError, f"{TESTFNnoat}.__all__ must be str"):
+                exec(f"from {TESTFNnoat} import *")
```

Comment:
Wouldn't be safer to pass locals and globals dicts? And you can check that the import doesn't have effect on them.